src/fipp/engine.cljc

```diff
@@ -5,6 +5,12 @@
             [fipp.deque :as deque]))
 
 
+(defn ^{:private true} unchunk [s]
```

Comment:
r.e. `{:private true}` - I'd personally just use `defn-` (or like you say, not worry about making it private), but I know that and the short form of metadata can rankle some people so played it safe :-). Will just go ahead and make it public.

r.e. the mapcat issue - I think the bug here is actually a little bit more subtle. I had seen that issue, and tried using the version of `join` that @ghadishayban includes in the comments there, but still exhausted the heap. I've narrowed it down to this:

```clojure
;; doesn't exhaust heap
(dorun (concat (range) (range)))
(dorun (concat (range) (range) (range)))
(dorun (apply concat [(range) (range)]))

;; exhausts heap
(dorun (apply concat [(range) (range) (range]))

;; doesn't exhaust heap
(dorun (apply concat (unchunk [(range) (range) (range)])))
```

Turns out it's because of sequence chunking - when the `zs` argument to concat is chunked (as it is when using `apply`), its entire first chunk will be retained in memory (and hence the fully realised range contained in that chunk will be retained too). Thanks to @bronsa for pointing this out to me.

```clojure
user=> (defn x [x & y] (println (class y)))
#'user/x
user=> (x 1 2 3 4)
clojure.lang.ArraySeq
nil
user=> (apply x [1 2 3 4])
clojure.lang.PersistentVector$ChunkedSeq
nil
user=> (apply x (unchunk [1 2 3 4]))
clojure.lang.Cons
```